dns-server/src/storage.rs

```diff
@@ -0,0 +1,899 @@
+        &'a self,
+        req_id: &'b str,
+        generation: u64,
+    ) -> Result<UpdateGuard<'a, 'b>, UpdateError> {
+        let mut update = self.updating.lock().await;
+        if let Some(ref update) = *update {
+            let elapsed =
+                chrono::Duration::from_std(update.start_instant.elapsed())
+                    .context("elapsed duration out of range")?;
+            return Err(UpdateError::UpdateInProgress {
+                start_time: update.start_time,
+                elapsed,
+                generation: update.generation,
+                req_id: update.req_id.clone(),
+            });
+        }
+
+        *update = Some(UpdateInfo {
+            start_time: chrono::Utc::now(),
+            start_instant: std::time::Instant::now(),
+            generation,
+            req_id: req_id.to_string(),
+        });
+
+        Ok(UpdateGuard { store: self, req_id, finished: false })
+    }
+
+    /// Updates to a new generation of DNS data
+    ///
+    /// See module-level documentation for constraints and design.
+    pub(crate) async fn dns_config_update(
+        &self,
+        config: &DnsConfigParams,
+        req_id: &str,
+    ) -> Result<(), UpdateError> {
+        let log = &self.log.new(o!(
+            "req_id" => req_id.to_owned(),
+            "new_generation" => config.generation
+        ));
+
+        // Lock out concurrent updates.  We must not return until we've released
+        // the "updating" lock!  (See UpdateGuard's `drop` impl.)
+        let update = self.begin_update(req_id, config.generation).await?;
+
+        info!(log, "attempting generation update");
+        let result = self.do_update(config).await;
+        match &result {
+            Ok(_) => info!(log, "updated generation"),
+            Err(error) => {
+                error!(log, "failed update"; "error_message" => #%error);
+            }
+        };
+
+        // Release our lock on concurrent update.
+        update.finish().await;
+
+        result
+    }
+
+    async fn do_update(
+        &self,
+        config: &DnsConfigParams,
+    ) -> Result<(), UpdateError> {
+        let log = &self.log;
+        let generation = config.generation;
+
+        // First, check if we're already at least as new as what's being
+        // requested.  Because we should have exclusive access to updates right
+        // now, it shouldn't be possible for this to change after we've checked
+        // it.
+        let old_config = self.read_config()?;
+        if old_config.generation > generation {
+            return Err(UpdateError::BadUpdateGeneration {
+                current_generation: old_config.generation,
+                attempted_generation: config.generation,
+            });
+        }
+        if old_config.generation == generation {
+            return Ok(());
+        }
+
+        // Prune any trees in the db that are newer than the current
+        // generation.
+        self.prune_newer(&old_config);
```

Comment:
Should this be possible given the check at line 373 above?

Reply:
This is intended to handle the case that we previously attempted to update to this generation but crashed while doing it, leaving some possibly incomplete trees around (and never having updated the config to reflect the new generation).  This approach wipes whatever we found and starts the update again.  I updated this comment to explain that case, since it's pretty non-obvious.